clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -229,6 +252,36 @@ public void authenticate() throws IOException {
         }
     }
 
+    @Override
+    public void reauthenticate(ReauthenticationContext reauthenticationContext) throws IOException {
+        Authenticator previousAuthenticator = Objects.requireNonNull(reauthenticationContext).previousAuthenticator();
+        if (!(previousAuthenticator instanceof SaslClientAuthenticator))
+            throw new IllegalArgumentException(
+                    "Invalid previous authenticator in client-side re-authentication context: " + previousAuthenticator);
+        apiVersionsResponse = ((SaslClientAuthenticator) previousAuthenticator).apiVersionsResponse;
```

Comment:
Also set state to `SEND_HANDSHAKE_REQUEST`

Reply:
👍 I created a `PROCESS_APIVERSIONS_RESPONSE` state where we process the response we got -- either from the server in the auth case or from the previous authenticator in the re-auth case.  This will be the initial state when re-authenticating.  I think this makes it clear what is going on.